src/transformers/generation_utils.py

```diff
@@ -1875,20 +1881,23 @@ def contrastive_search(
             model_kwargs["use_cache"] = True
             model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
 
-            # if the first step in the loop, encode all the prefix and obtain three parameters: (1) past_key_values; (2) last_hidden_states; (3) logit_for_next_step
-            if step_counter == 0:
-                # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save the `encoder_outputs`
+            # if the first step in the loop, encode all the prefix and obtain three parameters: (1) past_key_values;
+            # (2) last_hidden_states; (3) logit_for_next_step
+            if model_kwargs.get("past") is None:
```

Comment:
TF uses this pattern to determine whether it is on the first generation iteration or not, when cache is present. Since caching is mandatory for contrastive search, we can use it to determine whether we are in the first step without an extra variable.